datatypes/src/raster/grid.rs

```diff
@@ -0,0 +1,416 @@
+use std::{
+    fmt::Debug,
+    ops::{Add, Div, Mul, Rem, Sub},
+};
+
+use num_traits::{One, Zero};
+use serde::{Deserialize, Serialize};
+
+use crate::util::Result;
+
+use super::grid_bounds::GridBoundingBox;
+
+pub trait GridSize {
+    /// An array with one entry for each axis of the `GridRect`
+    type ShapeArray: AsRef<[usize]>;
+
+    /// The number of axis of the `GridRect`
+    const NDIM: usize;
+
+    // size per axis in [.., y, x] order
+    fn axis_size(&self) -> Self::ShapeArray;
+    /// Size of the x-axis
+    fn axis_size_x(&self) -> usize {
+        match *self.axis_size().as_ref() {
+            [] => 0,
+            [.., a] => a,
+        }
+    }
+
+    /// Size of the y-axis
+    fn axis_size_y(&self) -> usize {
+        match *self.axis_size().as_ref() {
+            [] => 0,
+            [_] => 1,
+            [.., b, _] => b,
+        }
+    }
+
+    /// The number of elements the `GridRect` contains. This is bounded by size(usize).
+    fn number_of_elements(&self) -> usize;
+}
+/// A n-dimensional rectangle defining the bounds of a grid
+pub trait GridBounds: GridSize {
+    /// An array with one entry for each axis of the `GridRect`
+    type IndexArray: AsRef<[isize]> + Into<GridIdx<Self::IndexArray>>;
+
+    /// The minimal valid index. This has the min valid value for each axis in the `GridRect`.
+    fn min_index(&self) -> GridIdx<Self::IndexArray>;
+    /// The max valid index. This has the max valid value for each axis in the `GridRect`.
+    fn max_index(&self) -> GridIdx<Self::IndexArray>;
+}
+
+pub trait BoundedGrid {
+    /// An array with one entry for each axis of the `GridRect`
+    type IndexArray: AsRef<[isize]> + Into<GridIdx<Self::IndexArray>>;
+    fn bounding_box(&self) -> GridBoundingBox<Self::IndexArray>;
+}
+
+pub trait GridContains<Rhs = Self> {
+    fn contains(&self, rhs: Rhs) -> bool;
+}
+
+pub trait GridIntersection<Rhs = Self, Out = Self> {
+    fn intersection(&self, other: &Rhs) -> Option<Out>;
+}
+
+pub trait GridSpaceToLinearSpace: GridSize {
+    /// An array with one entry for each axis of the `GridRect`
+    type IndexArray: AsRef<[isize]>;
+    /// The number of axis of the `GridRect`
+
+    /// Strides indicate how many linear space elements the next element at the same position of any axis is away.
+    fn strides(&self) -> Self::ShapeArray;
+    /// Calculate the zero based linear space location of an index in a `GridRect`.
+    fn linear_space_index_unchecked<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> usize;
+    /// Calculate the zero based linear space location of an index in a `GridRect`.
+    /// # Errors
+    /// This method fails if the grid index is out of bounds.
+    fn linear_space_index<I: Into<GridIdx<Self::IndexArray>>>(&self, index: I) -> Result<usize>;
+}
+
+pub trait GridIndexAccessMut<T, I> {
```

Comment:
should maybe inherit non mutable access?

Reply:
there is no dependency so this would be an unnecessary bound.